<a href="https://colab.research.google.com/github/bjentwistle/PythonFundamentals/blob/main/Worksheets/Numpy_mini_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Investigate, wrangle, add new data to the dataframe, filter and correlate

The following data file contains data about student scores in math, reading and writing.  The exam scores are assumed to be percentages. It also contains data about gender, ethnicity, parental education, whether the student qualifies for free school food and whether or not the student has taken a preparation course for the exams.  The data set has already been cleaned. 

1.  Investigate the data set.

2.  Create numpy arrays to hold each of the three sets of scores.  
Create a new numpy array to hold the average exam score (of the three scores for each student)
Add the new numpy array as a new column in the dataframe.

3.  Filter the original dataset into a new dataframe containing just the females.  Calculate the average exam scores for all rows in this new dataframe and then find the mean of the average column.

4.  Do the same for the males.

5.  Use the original dataset to find the correlation coefficient for reading and maths.  How closely do they correlate?  Write what you find in a text box below the code.

6.  Do the same for reading and writing. What do you find?

7.  You might want to filter on different criteria and check correlation (e.g. those on free school meals, or those who had prepared, etc)







The dataset can be accessed here:  https://raw.githubusercontent.com/lilaceri/Working-with-data-/main/Data%20Sets%20for%20code%20divisio/exams.csv.  This is a .csv file.

**NOTE:** Some useful references are included at the bottom of this spreadsheet.

Use the code cell below to work your code.

In [13]:
import pandas as pd
import numpy as np

url = "https://raw.githubusercontent.com/lilaceri/Working-with-data-/main/Data%20Sets%20for%20code%20divisio/exams.csv"
df = pd.read_csv(url)


def get_student():
  math_score = df['math_score'].to_numpy(np.float16)
  reading_score = df['reading_score'].to_numpy(np.float16)
  writing_score = df['writing_score'].to_numpy(np.float16)

  sum_of_scores = (math_score + reading_score + writing_score)
  mean_of_scores = sum_of_scores/3
  df['mean of scores'] = mean_of_scores.tolist()

  return df

df = get_student() #assigning the return value of the function 

#filter for female results
def find_female_results():
  female_filter = df[df['gender'] == "female"] #filtering on female only results
  
  #converting the results columns into numpy arrays to allow manipulation
  math_score = female_filter['math_score'].to_numpy(np.float16) 
  reading_score = female_filter['reading_score'].to_numpy(np.float16)
  writing_score = female_filter['writing_score'].to_numpy(np.float16)

  sum_of_scores = (math_score + reading_score + writing_score) #using numpy powers to add values in the 3 arrays
  mean_of_scores = sum_of_scores/3 #creating another array as the mean of the scores in each row.
  mean_of_faverages = mean_of_scores.mean() #working out the overall average for the female scores

  female_filter['mean of scores'] = mean_of_scores.tolist() #putting the column 'mean_of_scores' back on the panda df

  return mean_of_faverages #returning the female overall average score

mean_of_faverages = find_female_results() #assigning the return value of the function 

print("Mean of female scores = ", mean_of_faverages)

#filter for male results
def find_male_results():
  male_filter = df[df['gender']== "male"] #filtering on male only results

  math_score = male_filter['math_score'].to_numpy(np.float16)
  reading_score = male_filter['reading_score'].to_numpy(np.float16)
  writing_score = male_filter['writing_score'].to_numpy(np.float16)

  sum_of_scores = (math_score + reading_score + writing_score)
  mean_of_scores = sum_of_scores/3
  mean_of_maverages = mean_of_scores.mean()

  male_filter['mean of scores'] = mean_of_scores.tolist() #putting the column 'mean_of_scores' back on the panda df

  return mean_of_maverages #returning the male overall average score

mean_of_maverages = find_male_results() #assigning the return value of the function 
print("Mean of male scores = ", mean_of_maverages)

Mean of female scores =  70.8
Mean of male scores =  66.0


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:34: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:54: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [10]:
def get_correlation():
  # add your code below to get the correlation figure for maths and reading results
  #creating numpy lists from df columns reading_score and math_score
  reading = df["reading_score"].to_numpy(np.float16)
  maths = df['math_score'].to_numpy(np.int8)

  corr_coeff = np.corrcoef(reading, maths)
  return corr_coeff

corr_coeff = get_correlation()
print(corr_coeff[0][1])

0.8145975152425367


**My Findings**
The Maths and Reading scores do correlate a lot, 0.815 is a close correlation coefficient.

This suggests that maths and reading scores increase together (the correlation is postive so as one increases the other also increases) and that the correlation is quite strong.



In [11]:
def get_correlation_randw():
  # add your code below to get the correlation figure for maths and reading results
  #creating numpy lists from df columns reading_score and math_score
  reading = df["reading_score"].to_numpy(np.float16)
  writing = df['writing_score'].to_numpy(np.int8)

  corr_coeff_randw = np.corrcoef(reading, writing)
  return corr_coeff_randw

corr_coeff_randw = get_correlation_randw()
print(corr_coeff_randw[0][1])

0.9567686336066267


**My Findings**
There is an even stronger correlation (0.957) between reading and writing scores than the maths and reading. I personally,as a dyslexic student, can understand that those good at maths might not be as good at reading. I also struggle with reading more than writing. 

### Helpful references
---

Filtering on criteria using df references:  
`filtered_df = df[df['column name' == value]]`  

Filtering on multiple criteria using df reference:  
`filtered_df = df[df['first column name' == value] & df[second column name] >= value]`

Numpy help sheet:  http://datacamp-community-prod.s3.amazonaws.com/da466534-51fe-4c6d-b0cb-154f4782eb54

